cime/src/share/util/shr_reprosum_mod.F90

```diff
@@ -396,38 +433,81 @@ subroutine shr_reprosum_calc (arr, arr_gsum, nsummands, dsummands,     &
 !
 !-----------------------------------------------------------------------
 !
-! check whether input contains NaNs or INFs, and abort if so
+! initialize local statistics variables
+      gbl_lor_red = 0
+      gbl_max_red = 0
+      repro_sum_fast = 0
+      repro_sum_slow = 0
+      repro_sum_both = 0
+      nonrepro_sum = 0
 
-      call t_startf('shr_reprosum_NaN_INF_Chk')
-      nan_check = .false.
-      inf_check = .false.
-      num_nans = 0
-      num_infs = 0
+! set MPI communicator
+      if ( present(commid) ) then
+         mpi_comm = commid
+      else
+         mpi_comm = MPI_COMM_WORLD
+      endif
+      call t_barrierf('sync_repro_sum',mpi_comm)
 
-      nan_check = any(shr_infnan_isnan(arr))
-      inf_check = any(shr_infnan_isinf(arr))
-      if (nan_check .or. inf_check) then
-         do ifld=1,nflds
-            do isum=1,nsummands
-               if (shr_infnan_isnan(arr(isum,ifld))) then
-                  num_nans = num_nans + 1
-               endif
-               if (shr_infnan_isinf(arr(isum,ifld))) then
-                  num_infs = num_infs + 1
-               endif
-            end do
-         end do
+! check whether should abort if input contains NaNs or INFs
+      abort_inf_nan = .not. repro_sum_allow_infnan
+      if ( present(allow_infnan) ) then
+         abort_inf_nan = .not. allow_infnan
       endif
-      call t_stopf('shr_reprosum_NaN_INF_Chk')
 
-      if ((num_nans > 0) .or. (num_infs > 0)) then
-         call mpi_comm_rank(MPI_COMM_WORLD, mypid, ierr)
-         write(s_logunit,37) real(num_nans,r8), real(num_infs,r8), mypid
+      call t_startf('shr_reprosum_INF_NaN_Chk')
+
+! initialize flags to indicate that no NaNs or INFs are present in the input data
+      inf_nan_gchecks = .false.
+      arr_gsum_infnan = .false.
+
+      if (abort_inf_nan) then
+
+! check whether input contains NaNs or INFs, and abort if so
+         nan_check = any(shr_infnan_isnan(arr))
+         inf_check = any(shr_infnan_isinf(arr))
+
+         if (nan_check .or. inf_check) then
+
+            nan_count = count(shr_infnan_isnan(arr))
+            inf_count = count(shr_infnan_isinf(arr))
```

Comment:
The calls to `count()` make the calls to `any()` above redundant (second array traversal): remove calls to `any()` and the `if (nan_check .or. inf_check)`?

Reply:
Depends on whether any() is faster than count(). If any() is .true., then we don't care about performance at this fine grain (since we will be aborting). Trying to optimizing for the case where we care about performance.